### DISCUSSION TITLE
Using groupBy without selecting the PK

### DISCUSSION THREAD

**Discussion by co-sic** [mikro-orm/mikro-orm#3424] [Q&A] [answered]
I have an Entity with a unique id and a non unique field. I want to do a query grouping by the non unique field and getting the values for that field back (I'm adding also adding a where query and orderBy, but that should not be relevant i think).
But if i use the groupBy clause from the em find method, i get an error `column \"p0.id\" must appear in the GROUP BY clause or be used in an aggregate function` which makes sense. I don't need the id field in my result, but if i use the fields options id is always auto selected. I guess this is always needed for the IdentityMap.
I tried using the querybuild of the em, but then i get another error `You cannot merge entity '...' without identifier!`, i guess this is also due to the id field not being selected.

So lets say i have some Book entity like below:

| id | year |
| - | - |
| 1 | 2001 |
| 2 | 2001 |
| 3 | 2001 |
| 4 | 1990 |
| 5 | 1950 |

I want this as a result:

|  year |
| - |
| 2001 |
| 1990 |
| 1950 |

Any pointers?

**Comment by B4nan** [comment, marked as answer]
> I tried using the querybuild of the em, but then i get another error You cannot merge entity '...' without identifier!, i guess this is also due to the id field not being selected.

You are probably using `qb.getResult()` or awaiting the QB instance. That means mapping to entities, and that requires the PK to be there. You can use `qb.execute()` to get the raw results instead.

**Reply by co-sic** [reply]
Thanks for the quick reply, that works :+1: 

But is there a way to achieve the same result with the find method of the em? I'm wondering what a use case would be for the groupBy option, since you could only group by the primary key as far as i can tell.

**Reply by B4nan** [reply]
Nope, that method will always give you managed entity, and such needs to have a PK. Given your OP it sounds like you dont want to return entities at all, rather an array of numbers (or objects with a single key), those are not entities, so you should not try to get such from _Entity_Manager :]
